adapters/criteo/models.go

```diff
@@ -0,0 +1,333 @@
+
+	criteoRequest.User = newCriteoUser(request.User, request.Device, regsExt)
+
+	if gdprConsent, err := newCriteoGdprConsent(request.User, regsExt); err != nil {
+		errs = append(errs, err)
+	} else {
+		criteoRequest.GdprConsent = gdprConsent
+	}
+
+	if request.User != nil && request.User.Ext != nil {
+		var extUser openrtb_ext.ExtUser
+		if err := json.Unmarshal(request.User.Ext, &extUser); err != nil {
+			errs = append(errs, err)
+		} else {
+			criteoRequest.Eids = extUser.Eids
+		}
+	}
+
+	return criteoRequest, errs
+}
+
+type criteoPublisher struct {
+	SiteID    string `json:"siteid,omitempty"` // TODO: make sure it's siteid and not publisherid
+	BundleID  string `json:"bundleid,omitempty"`
+	URL       string `json:"url,omitempty"`
+	NetworkID *int64 `json:"networkid,omitempty"`
+}
+
+func newCriteoPublisher(networkId *int64, app *openrtb2.App, site *openrtb2.Site) criteoPublisher {
+	// Both app and site cannot be nil at the same time by design in PBS
+
+	criteoPublisher := criteoPublisher{}
+
+	if networkId != nil && *networkId > 0 {
+		criteoPublisher.NetworkID = networkId
+	}
+
+	if app != nil {
+		if app.Bundle != "" {
+			criteoPublisher.BundleID = app.Bundle
+		}
+	}
+
+	if site != nil {
+		if site.ID != "" {
+			criteoPublisher.SiteID = site.ID
+		}
+		if site.Page != "" {
+			criteoPublisher.URL = site.Page
+		}
+	}
+
+	return criteoPublisher
+}
+
+type criteoUser struct {
+	DeviceID     string `json:"deviceid,omitempty"`
+	DeviceOS     string `json:"deviceos,omitempty"`
+	DeviceIDType string `json:"deviceidtype,omitempty"`
+	CookieID     string `json:"cookieuid,omitempty"`
+	UID          string `json:"uid,omitempty"`
+	IP           string `json:"ip,omitempty"`
+	UA           string `json:"ua,omitempty"`
+	UspIab       string `json:"uspIab,omitempty"`
+}
+
+func newCriteoUser(user *openrtb2.User, device *openrtb2.Device, regsExt *openrtb_ext.ExtRegs) criteoUser {
+	criteoUser := criteoUser{}
+
+	if user == nil && device == nil {
+		return criteoUser
+	}
+
+	if user != nil {
+		if user.BuyerUID != "" {
+			criteoUser.CookieID = user.BuyerUID
+		}
+	}
+
+	if device != nil {
+		deviceType := getDeviceType(device.OS)
+		criteoUser.DeviceIDType = deviceType
+
+		if device.OS != "" {
+			criteoUser.DeviceOS = device.OS
+		}
+		if device.IFA != "" {
+			criteoUser.DeviceID = device.IFA
+		}
+		if device.IP != "" {
+			criteoUser.IP = device.IP
+		}
+		if device.UA != "" {
+			criteoUser.UA = device.UA
+		}
+	}
+
+	if regsExt != nil {
+		if regsExt.USPrivacy != "" {
+			criteoUser.UspIab = regsExt.USPrivacy // CCPA
+		}
+	}
+
+	return criteoUser
+}
+
+type criteoGdprConsent struct {
+	GdprApplies *bool  `json:"gdprapplies,omitempty"`
+	ConsentData string `json:"consentdata,omitempty"`
+}
+
+func newCriteoGdprConsent(user *openrtb2.User, regsExt *openrtb_ext.ExtRegs) (criteoGdprConsent, error) {
+	consent := criteoGdprConsent{}
+
+	if user == nil && regsExt == nil {
+		return consent, nil
+	}
+
+	if user != nil {
+		if user.Ext != nil {
```

Comment:
Please consider combining these if statements to a single line, as you have in other funcs in this file.
`if user != nil && user.Ext != nil {` ...

Reply:
Ack